db/db.ts

```diff
@@ -310,3 +317,110 @@ export const getHomepageId = (
             AND published = TRUE`
     ).then((result) => result?.id)
 }
+
+export const getImageMetadataByFilenames = async (
+    knex: KnexReadonlyTransaction,
+    filenames: string[]
+): Promise<Record<string, ImageMetadata & { id: number }>> => {
+    if (filenames.length === 0) return {}
+    const rows = await knexRaw<ImageMetadata & { id: number }>(
+        knex,
+        `-- sql
+        SELECT
+            id,
+            googleId,
+            filename,
+            defaultAlt,
+            updatedAt,
+            originalWidth,
+            originalHeight
+        FROM
+            images
+        WHERE filename IN (?)`,
+        [filenames]
+    )
+    return keyBy(rows, "filename")
+}
+
+export const getPublishedGdocPosts = async (
+    knex: KnexReadonlyTransaction
+): Promise<DbEnrichedPostGdoc[]> => {
+    return knexRaw<DbRawPostGdoc>(
+        knex,
+        `-- sql
+        SELECT
+        g.breadcrumbs,
+        g.content,
+        g.createdAt,
+        g.id,
+        g.markdown,
+        g.publicationContext,
+        g.published,
+        g.publishedAt,
+        g.revisionId,
+        g.slug,
+        g.updatedAt
+    FROM
+        posts_gdocs g
+    WHERE
+        g.published = 1
+        AND g.content ->> '$.type' IN (:types)
+        AND g.publishedAt <= NOW()
+    ORDER BY g.publishedAt DESC`,
+        {
+            types: [
+                OwidGdocType.Article,
+                OwidGdocType.LinearTopicPage,
+                OwidGdocType.TopicPage,
+                OwidGdocType.AboutPage,
+            ],
+        }
+    ).then((rows) => rows.map(parsePostsGdocsRow))
+}
+
+export const getPublishedGdocPostsWithTags = async (
+    knex: KnexReadonlyTransaction
+): Promise<DBEnrichedPostGdocWithTags[]> => {
+    return knexRaw<DBRawPostGdocWithTags>(
+        knex,
+        `-- sql
+        SELECT
+        g.breadcrumbs,
+        g.content,
+        g.createdAt,
+        g.id,
+        g.markdown,
+        g.publicationContext,
+        g.published,
+        g.publishedAt,
+        g.revisionId,
+        g.slug,
+        g.updatedAt,
+        JSON_ARRAYAGG(
+          JSON_OBJECT(
+            'id', t.id,
+            'name', t.name,
+            'slug', t.slug
+        )) AS tags
+    FROM
+        posts_gdocs g
+    JOIN posts_gdocs_x_tags gxt ON
+        g.id = gxt.gdocId 
+    JOIN tags t ON
+        gxt.tagId = t.id
```

Comment:
If the goal is to also include gdocs without tags - which I think it is - then these need to be left joins:
```suggestion
    LEFT JOIN posts_gdocs_x_tags gxt ON
        g.id = gxt.gdocId 
    LETT JOIN tags t ON
        gxt.tagId = t.id
```

Reply:
Oops, misremembered what the default join is. That explains why my test output was flawless 😁 

Thanks friends!